src/main/scala/is/hail/io/bgen/BgenWriter.scala

```diff
@@ -0,0 +1,290 @@
+      index(i) = i
+      i += 1
+    }
+  }
+
+  def resizeWeights(input: Array[Double], output: ArrayBuilder[Int], resized: Array[Double], index: Array[Int], size: Long) {
+    val n = input.length
+    assert(resized.length == n && index.length == n)
+    resetIndex(index)
+
+    val conversionFactor = size.toDouble / input.sum
+
+    var totalFractional = 0d
+    var i = 0
+    while (i < n) {
+      val x = input(i) * conversionFactor
+      resized(i) = x
+      totalFractional += (x - x.floor)
+      i += 1
+    }
+
+    val F = (totalFractional + 0.5).toInt
+    assert(F >= 0 && F <= n)
+
+    sortedIndex(resized, index)
+
+    i = 0
+    var newSize = 0d
+    while (i < n) {
+      val r = if (index(i) < F) resized(i).ceil else resized(i).floor
+      if (i != n - 1)
+        output += r.toUInt.intRep
+      newSize += r
+      i += 1
+    }
+
+    assert(newSize == size)
+  }
+
+  def resizeWeights(input: Array[Int], output: ArrayBuilder[Int], resized: Array[Double], index: Array[Int], size: Long) {
+    val n = input.length
+    assert(resized.length == n && index.length == n)
+    resetIndex(index)
+
+    val conversionFactor = size.toDouble / input.sum
+
+    var totalFractional = 0d
+    var i = 0
+    while (i < n) {
+      val x = input(i).toUIntFromRep * conversionFactor
+      resized(i) = x
+      totalFractional += (x - x.floor)
+      i += 1
+    }
+
+    val F = (totalFractional + 0.5).toInt
+    assert(F >= 0 && F <= n)
+
+    sortedIndex(resized, index)
+
+    i = 0
+    var newSize = 0d
+    while (i < n) {
+      val r = if (index(i) < F) resized(i).ceil else resized(i).floor
+      if (i != n - 1)
+        output += r.toUInt.intRep
+      newSize += r
+      i += 1
+    }
+
+    assert(newSize == size)
+  }
+
+  def emitHeaderBlock(vds: VariantDataset): Array[Byte] = {
+    val nSamples = vds.nSamples
+    val sampleIds = vds.stringSampleIds.toArray
+    val nVariants = vds.countVariants()
+    assert(nVariants <= (1L << 32) - 1, s"Maximum number of variants can export is (2^32 - 1). Found $nVariants.")
+
+    val magicNumbers = Array("b", "g", "e", "n").flatMap(_.getBytes)
+    val flags = 0x01 | (0x02 << 2) | (0x01 << 31)
+    val headerLength = 20
+
+    val bb = new ArrayBuilder[Byte]
+
+    intToBytesLE(bb, 0) // placeholder for offset
+    intToBytesLE(bb, headerLength)
+    intToBytesLE(bb, nVariants.toUInt.intRep)
+    intToBytesLE(bb, nSamples)
+    bb ++= magicNumbers
+    intToBytesLE(bb, flags)
+
+    intToBytesLE(bb, 0) // placeholder for length of sample ids
+    intToBytesLE(bb, nSamples)
+
+    var i = 0
+    var sampleBlockLength = 8
+    while (i < nSamples) {
+      sampleBlockLength += stringToBytesWShortLength(bb, sampleIds(i))
+      i += 1
+    }
+
+    val offset = headerLength + sampleBlockLength
+    updateIntToBytesLE(bb, offset, 0)
+    updateIntToBytesLE(bb, sampleBlockLength, 24)
+
+    bb.result()
+  }
+
+  def emitVariantBlock(v: Variant, va: Annotation, gs: Iterable[Genotype],
```

Comment:
Actually, leave the emit, but this should take a ArrayBuilder which is allocated per-partition.  See comment below.